src/Data/Singletons/Single/Fixity.hs

```diff
@@ -7,21 +7,42 @@ import Data.Singletons.Util
 import Data.Singletons.Names
 import Language.Haskell.TH.Desugar
 
-singInfixDecl :: Fixity -> Name -> DLetDec
+singInfixDecl :: DsMonad q => Fixity -> Name -> q DLetDec
 singInfixDecl fixity name
-  | isUpcase name =
-    -- is it a tycon name or a datacon name??
-    -- it *must* be a datacon name, because symbolic tycons
-    -- can't be promoted. This is terrible.
-    DInfixD fixity (singDataConName name)
-  | otherwise = DInfixD fixity (singValName name)
+  | data_con_name
+  , ':':_ <- nameBase name
+    -- We must be careful with infix constructor names here. It's not
+    -- sufficient to check if a name begins with a colon to conclude that it
+    -- is in fact the name of a constructor. For instance, :*: could name a
```

Comment:
It turns out `reifyNameSpace` needs to be monadic to work properly in all cases, but regardless, `reifyNameSpace` made the implementation of `singInfixDecl` _much_ cleaner. (See https://github.com/goldfirere/th-desugar/pull/64 for the `th-desugar` PR for `reifyNameSpace`.)